src/address_space.rs

```diff
@@ -34,61 +57,156 @@ impl AddressSpace {
+            return Ok(mapping_addr);
+        }
+        Err("out of address space!")
     }
 
     /// Add a mapping from `DataSource` into this `AddressSpace` starting at a specific address.
     ///
     /// # Errors
     /// If there is insufficient room subsequent to `start`.
-    pub fn add_mapping_at<D: DataSource>(
-        &self,
-        source: &D,
+    pub fn add_mapping_at<D: DataSource + 'static>(
+        &mut self,
+        source: Arc<D>,
         offset: usize,
         span: usize,
         start: VirtualAddress,
+        flags: FlagBuilder,
     ) -> Result<(), &str> {
-        todo!()
+        let adjusted_span: usize = span + PAGE_SIZE - (span % PAGE_SIZE);
+        let end = start + adjusted_span;
+        for mapping in &self.mappings {
+            if mapping.addr + mapping.span < start {
+                // we have not reached the area we need yet
+                continue;
+            } else if start + adjusted_span < mapping.addr {
+                // we've moved past the area we need
+                break;
+            } else if start <= mapping.addr && mapping.addr <= start + adjusted_span {
+                // there is a mapping that starts in the space we want to place the new mapping
+                return Err("out of address space!");
+            } else if start <= mapping.addr + mapping.span
+                && mapping.addr + mapping.span <= start + adjusted_span
+            {
+                // there is a mapping that ends in the space we want to place the new mapping
+                return Err("out of address space!");
+            } else if mapping.addr < start && start + adjusted_span < mapping.addr + mapping.span {
+                // there is a mapping covers the space we want to place the new mapping
+                return Err("out of address space!");
+            }
+        }
+        // if we have enough space for our new mapping
+        let new_mapping = MapEntry::new(source.clone(), offset, adjusted_span, start, flags);
+        self.mappings.push(new_mapping);
+        self.mappings.sort_by(|a, b| a.addr.cmp(&b.addr));
+        return Ok(());
     }
 
     /// Remove the mapping to `DataSource` that starts at the given address.
     ///
     /// # Errors
     /// If the mapping could not be removed.
     pub fn remove_mapping<D: DataSource>(
-        &self,
-        source: &D,
+        &mut self,
+        source: Arc<D>,
         start: VirtualAddress,
     ) -> Result<(), &str> {
-        todo!()
+        let mapping = self.get_mapping_for_addr(start).unwrap();
+        let s = &self.mappings.len();
+
+        if let Some(pos) = self.mappings.iter().position(|x| x.addr == mapping.addr) {
+            self.mappings.remove(pos);
+        } else {
+            return Err("error removing mapping");
+        }
+        if self.mappings.len() != s - 1 {
+            return Err("error removing mapping");
+        }
+        drop(source);
```

Comment:
This is unnecessary - it will be dropped automatically when it goes out of scope, i.e. at the end of the function.